### ISSUE TITLE
[Quantization] FullyConnected lowered into MatMul + BatchedAdd is not quantized

### ISSUE THREAD

**Issue by mciprian13** [pytorch/glow#4630]
Given the current model: [model.zip](https://github.com/pytorch/glow/files/4821032/model.zip)
which is similar to MobileNet and which ends with MatMul + Add like this:
![image](https://user-images.githubusercontent.com/29785908/85437569-78d0ef00-b593-11ea-9427-298027efa4f1.png)
When trying to quantize this model the following happens:
- in the folding stage the MatMul + Add is correctly folded into a FullyConnected
- in the lowering stage the FullyConnected is correctly lowered into MatMul + BatchedAdd
- the problem is that during the quantization step the MatMul and BatchedAdd are NOT quantized
- similar thing happens when a 1x1 Conv is lowered into FC and further into MatMul + BatchedAdd
- interestingly a FullyConnected node from a Caffe2 model gets quantized (after lowered into MatMul + BatchedAdd)

All the nodes in the graph are quantized except the MatMul and BatchedAdd:
![image](https://user-images.githubusercontent.com/29785908/85439422-ad45aa80-b595-11ea-8d17-ecd31588e278.png)

I used the following commands:
```
image-classifier ./images/*.png -model=model.onnx -model-input-name=input_1 -image-mode=neg1to1 -image-layout=NHWC -image-channel-order=RGB -dump-profile=profile.yml
model-compiler -backend=CPU -emit-bundle=bundle -model=model.onnx -load-profile=profile.yml -dump-graph-DAG=graph.dot
```
The images used for profiling are those in the Glow repository: https://github.com/pytorch/glow/tree/master/tests/images/imagenet

**Comment by mciprian13** [comment]
@jfix71 Can you help me with this? Give me a hint or something ... I spent a day on this and did not figured it out 😢

**Comment by jfix71** [comment]
@mciprian13 I looked a bit, and as far as I can something is happening where in the original model you have a `MatMul` `dense_1` followed by an `Add` that is broadcasted (*not* `BatchedAdd`). When I dump the graph using the `image-classifier` with or without profiling that is what I see, e.g.

<img width="400" alt="PM" src="https://user-images.githubusercontent.com/1198212/85640048-8479ef80-b63f-11ea-8f2d-7ae44049e41a.png">

I'm assuming this is because constant folding is kicking in somewhere (either during loading or during the optimization pipeline) which is converting a `Constant -> Tile -> Add` into `Constant' -> Add`. However what we want to happen is for the `FoldTileAddIntoBatchedAdd` opt to kick in before this happens to turn it into `Constant -> BatchedAdd`, which is likely what's happening when you run the `model-compiler`. I'm assuming then that there's no profile for these nodes when we run `image-classifier`, and so quantization doesn't occur when we load that profile and run it with `model-compiler`...?

My assumption might be wrong, it could also be due to some issue in the interplay of lowering/folding as you're somewhat suggesting. Regardless I'd suggest taking a look at the profile you dump to see if the nodes are there as expected. I'd also suggest trying e.g. `--dump-graph-before-all-passes`/`--dump-graph-after-all-passes` to try to examine what's going on optimization-by-optimization, e.g. in the `model-compiler` to see when it changes the Add into BatchedAdd and why that's not kicking in for `image-classifier`.

**Comment by mciprian13** [comment]
@jfix71 You assumption was correct. What happens here is that during profiling the model is compiled with batch size 3 resulting in a graph for which the optimization FoldMatMulAddIntoFullyConnected does not kick in (as it does during the actual quantization) because the optimization is limited for batch size 1. Thus during quantization some profiling info is not found.
A simple hack for this is to do the profiling with `-minibatch=1` resulting in the model being compiled with batch size 1 and run multiple times for different images.
I would suggest making the `-minibatch` default value equal to 1 since it might be that other optimizations might suffer from this as well. Best practice is to have the graph compiled in the same way during profiling/compiling. WDYT?

**Comment by jfix71** [comment]
>I would suggest making the -minibatch default value equal to 1 since it might be that other optimizations might suffer from this as well. Best practice is to have the graph compiled in the same way during profiling/compiling. WDYT?

Hmm, my main concern here is tooling/scripting we have that assumes current default behavior. Besides, even if it is set to default to 1, that doesn't really solve the problem in the generic case, right? I.e. you didn't specify the batch size for the `model-compiler` command via `-model-input`, so the batch size was probably just inferred from the model and the model has batch size 1, so it works in this case to use `-minibatch=1` but wouldn't in other cases.

One way to improve the situation at least would be to detect a mismatch on the graph that was used for generating the profile and the one that is then used later on that is quantized with that profile. Perhaps we could generate a hash of the graph after `optimizeFunctionBeforeLowering()` is called (e.g. [here](https://github.com/pytorch/glow/blob/b98b71210cdceece5d6c69536c9d0f281bf7717d/lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp#L4872)) and include that hash in the profile? Then if we could somehow save the hash in the profile and check its the same when its loaded for quantizing the model later on, if there's a mismatch it signals there was a difference in the graph between profiling and quantization. It would be a good debugging signal that something has gone wrong. WDYT?

**Comment by mciprian13** [comment]
@jfix71 This sound like a useful tool for debugging (and maybe deserves opening an issue for this) but still does not solve my problem. What I did I generalized the FoldMatMulAddIntoFullyConnected optimization to take care of multibatch and now the MatMul gets quantized but the BatchedAdd node still does not get quantized because during profiling the bias operand gets broadcasted and the constant folding changes the name of the constant node while during quantization no broadcasting occurs and so no profiling info is found for that node.
1. Don't you think it would be good to disable the constant folding optimization during profiling such that the profiling info is found during quantization (e.g. whether a constant is broacasted or not)?
2. Or maybe disable the constant folding for broadcasting operators only since only these are related to batching?
3. Or maybe fix the constant folding for broadcasting operators to maintain the name of the constant? Actually this is a good question: shouldn't the constant folding maintain the name of the initial constant?
WDYT?

**Comment by mciprian13** [comment]
@jfix71 I opened a related PR here #4636.

**Comment by jfix71** [comment]
>I checked and I see no hashing method for Function/Module

One easy way to do this is to hash `Function::toString()`.

RE: Moving to `minibatch=1` default, I'm generally fine with it. But we need to do this carefully to make sure scripts/tooling that assumes current default behavior is updated. and make sure it's broadcasted clearly so that other users are aware of this change.

RE: constant folding and profiling/quantization -- we need the Function to be the exact same up until we get to lowering, i.e. the place I linked to in my previous comment. This means if we constant fold before quantization we need to also constant fold before profiling and vice versa. One potential option could be to delay constant folding until after quantization -- though this also will change numerics 

Are your suggestions for disabling constant folding under the assumption that the batch size can differ between profiling and quantization? If so, I think that path is going to be hard to support well.

I guess I'm wondering if there's still a problem if we make sure users use the same batch size for profiling and quantization -- and doing e.g. the hash checking thing is a good way to flag this sort of problem quickly/easily with an error message suggesting potential reasons this occurred (e.g. this could also be because the profile was created from an older version of the compiler that had different optimizations enabled, used different name uniquing, etc.).

**Comment by mciprian13** [comment]
@jfix71 The idea with the hash is very good. I`m going to do it. A couple of questions:
- do you know a function utility to hash a string?
- the returned hash type should be llvm::hash_code (just as for the Node)?
After that I`ll have to figure out how to print the hash in the profile but it should be doable.

**Comment by mciprian13** [comment]
@jfix71 I made in #4636 also the hash checking mechanism. Have a look!

**Comment by mciprian13** [comment]
Solved in #4636. Closing this issue.
